src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedObjectCreationTests.cs

```diff
@@ -0,0 +1,3926 @@
+    }
+}
+";
+
+            var comp = CreateCompilation(source, options: TestOptions.DebugExe).VerifyDiagnostics();
+
+            CompileAndVerify(comp, expectedOutput: "C");
+        }
+
+        [Fact]
+        public void TestInNullCoalescingAssignment()
+        {
+            var source = @"
+using System;
+
+class C
+{
+    public static void M<T>(Func<bool, T> f)
+    {
+        Console.Write(f(true));
+        Console.Write(f(false));
+    }
+    public static void Main()
+    {
+        C x = null;
+        x ??= new();
+        Console.Write(x);
+    }
+}
+";
+
+            var comp = CreateCompilation(source, options: TestOptions.DebugExe).VerifyDiagnostics();
+
+            CompileAndVerify(comp, expectedOutput: "C");
+        }
+
+        [Fact]
+        public void TestInNullCoalescingAssignment_ErrorCase()
+        {
+            var source = @"
+class C
+{
+    public static void Main()
+    {
+        new() ??= new C();
+        new() ??= new();
+    }
+}
+";
+
+            var comp = CreateCompilation(source, options: TestOptions.DebugExe).VerifyDiagnostics(
+                // (6,9): error CS0131: The left-hand side of an assignment must be a variable, property or indexer
+                //         new() ??= new C();
+                Diagnostic(ErrorCode.ERR_AssgLvalueExpected, "new()").WithLocation(6, 9),
+                // (7,9): error CS0131: The left-hand side of an assignment must be a variable, property or indexer
+                //         new() ??= new();
+                Diagnostic(ErrorCode.ERR_AssgLvalueExpected, "new()").WithLocation(7, 9)
+                );
+        }
+
+        [Fact]
+        public void TestBestType_Lambda_ErrorCase()
+        {
+            var source = @"
+using System;
+
+class C
+{
+    public static void M<T>(Func<bool, T> f)
+    {
+    }
+
+    public static void Main()
+    {
+        M(b => { if (b) return new(); else return new(); });
+    }
+}
+";
+
+            var comp = CreateCompilation(source, options: TestOptions.DebugExe).VerifyDiagnostics(
+                // (12,9): error CS0411: The type arguments for method 'C.M<T>(Func<bool, T>)' cannot be inferred from the usage. Try specifying the type arguments explicitly.
+                //         M(b => { if (b) return new(); else return new(); });
+                Diagnostic(ErrorCode.ERR_CantInferMethTypeArgs, "M").WithArguments("C.M<T>(System.Func<bool, T>)").WithLocation(12, 9)
+                );
+        }
+
+        [Fact]
+        public void TestBadTypeParameter()
+        {
+            var source = @"
+class C
+{
+    static void M<A, B, C>()
+        where A : struct
+        where B : new()
+    {
+        A v1 = new(1);
+        B v2 = new(2);
+        C v3 = new();
+    }
+}
+";
+
+            var comp = CreateCompilation(source, options: TestOptions.ReleaseDll).VerifyDiagnostics(
+                // (8,16): error CS0417: 'A': cannot provide arguments when creating an instance of a variable type
```

Comment:
Is this missing some of the error message?

Reply:
What error message are you referring to?

Reply:
I'm just not sure what this error message is saying in general. What does a `variable type` mean? Perhaps it would be better to say `type argument?

Reply:
We produce the same existing error from `new T(0)`. I can change it here but I don't think it's particularly related to this PR.

Reply:
Alright, that's fine then.